### PR TITLE
Fix issues in with keyword implementation

### PR DESCRIPTION
**Pull request by tsandall**
This PR fixes issues #1261 and #802 as well as a previously unknown problem (commit bd690c1) in the with keyword implementation. There is also a bit of refactoring in the PR that makes the with keyword tests easier to maintain.

### CODE REVIEW

**Comment by tsandall** [comment]
Rebased to fix missed DCO sign-off.

File: topdown/cache.go [hunk 1]
```diff
@@ -171,3 +171,36 @@ func (e *baseCacheElem) set(value ast.Value) {
...
+
+func (s *refStack) Push(refs []ast.Ref) {
+	s.sl = append(s.sl, refStackElem{refs: refs})
+}
+
+func (s *refStack) Pop() {
+	s.sl = s.sl[:len(s.sl)-1]
+}
+
+func (s *refStack) Prefixed(ref ast.Ref) bool {
```
Comment (ashutosh-narkar) [inline topdown/cache.go]: Are there any performance implications of replacing the trie implementation ?

File: topdown/topdown_test.go [hunk 2]
```diff
@@ -2637,12 +2638,65 @@ func TestTopDownWithKeyword(t *testing.T) {
...
+			rules: []string{
+				`q = "b" { true }`,
+				`p = [x ,y] {
+					q = x with q as "a"
+					q = y
+				}`,
+			},
+		},
+		{
+			note: "with invalidate data stack iteration",
```
Comment (ashutosh-narkar) [inline topdown/topdown_test.go]: What is the `exp` result ?

**Comment by ashutosh-narkar** [comment]
[Data Mocking](https://www.openpolicyagent.org/docs/v0.10.7/how-do-i-test-policies/#data-mocking) section of the docs needs to be updated. For example, the compiler error message is outdated.

File: topdown/cache.go (hunk 1 above)
Comment (tsandall) [reply topdown/cache.go]: The check will be O(number of with targets in-scope) so I'm not too worried about it. If it becomes a problem we can revisit ti.

File: topdown/topdown_test.go (hunk 2 above)
Comment (tsandall) [reply topdown/topdown_test.go]: Nice catch! I've fixed this and improved the test helper to error if an unexpected expected value is received (in this case, the helper was receiving `nil` for the expected value which just fell through the switch statement).

**Pull request merged**